# Tarefa 1 - Instalacao do PostgreSQL

In [ ]:
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start
# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"

# Setup a database with name `sampledb` to be used
!sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sampledb;'
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sampledb;'

debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 12.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package cron.
(Reading database ... 155676 files and directories currently installed.)
Preparing to unpack .../00-cron_3.0pl1-128.1ubuntu1.2_amd64.deb ...
Unpacking cron (3.0pl1-128.1ubuntu1.2) ...
Selecting previously unselected package logrotate.
Preparing to unpack .../01-logrotate_3.11.0-0.1ubuntu1_amd64.deb ...
Unpacking logrotate (3.11.0-0.1ubuntu1) ...
Selecting previously unselected package netbase.
Preparing to unpack .../02-netbase_5.4_all.deb ...
Unpacking netbase (5.4) ...
Preparing to unpack .../03-l

In [ ]:
# set connection
%env DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sampledb

#To load the sql extention to start using %%sql
%load_ext sql

env: DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sampledb


In [ ]:
pip install db-sqlite3

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 171 kB 31.1 MB/s 
  Created wheel for db-sqlite3: filename=db_sqlite3-0.0.1-py3-none-any.whl size=1794 sha256=a8026ea2c1f159c34b203e9e10584f3676322ca67ef985808867e2175715ae07
  Stored in directory: /root/.cache/pip/wheels/02/38/d5/2f54461050571bf5330fee2a37ab1c9b5e7540b0572f1acdab
  Created wheel for db: filename=db-0.1.1-py3-none-any.whl size=3895 sha256=b564d8d0bdac97b4901b85203c8b09dd0196bfffb18fb463c379ea8173019364
  Stored in directory: /root/.cache/pip/wheels/8e/97/82/741d2b360507411ec233d0280d7371faa94b03bde834e4a9be
  Created wheel for antiorm: filename=antiorm-1.2.1-py3-none-any.whl size=31679 sha256=8011526174d2d599b1d58033da10e157c7159093c92e80cca1e9254c9fb519f3
  Stored in directory: /root/.cache/pip/wheels/c5/43/70/e9729370cfff40c49d3e3d05377d54b3ecd71f64e62341ea80
Successfully built db-sqlite3 db antiorm


In [ ]:
# Programa que testa a execução da biblioteca.
# Após o teste, conseguimos ver a versão atual do sqlite3 instalado: 2.6

import sqlite3
sqlite3.version

'2.6.0'

In [ ]:
# Também criamos uma tabela para testar a conexão, e conseguimos executar sem nenhum problema

conn = sqlite3.connect('test.db')

conn.execute('''
CREATE TABLE Departamentos (
   Code INTEGER PRIMARY KEY NOT NULL,
   Name NVARCHAR NOT NULL ,
   Budget REAL NOT NULL 
 );''')
 
conn.commit()
 
print("Departamentos table created");

Departamentos table created


# Tarefa 2 – Geração de um BD de testes


In [ ]:
!wget -q https://github.com/electrum/tpch-dbgen/archive/32f1c1b92d1664dba542e927d23d86ffa57aa253.zip -O tpch-dbgen.zip

In [ ]:
!unzip -q tpch-dbgen.zip && mv tpch-dbgen-32f1c1b92d1664dba542e927d23d86ffa57aa253 tpch-dbgen && rm tpch-dbgen.zip

In [ ]:
!pip3 install psycopg2-binary
!pip3 install mock

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 3.0 MB 8.3 MB/s 


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!cd tpch-dbgen

In [ ]:
!git clone https://github.com/foliveirafilho/tpch-pgsql

Cloning into 'tpch-pgsql'...
remote: Enumerating objects: 164, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 164 (delta 0), reused 0 (delta 0), pack-reused 160
Receiving objects: 100% (164/164), 774.38 KiB | 2.51 MiB/s, done.
Resolving deltas: 100% (69/69), done.


In [ ]:
cd tpch-pgsql/

/content/tpch-pgsql


In [ ]:
!python3 tpch_pgsql.py

usage: tpch_pgsql.py [-h] [-H HOST] [-p PORT] [-U USERNAME] [-W [PASSWORD]]
                     [-d DBNAME] [-i DATA_DIR] [-q QUERY_ROOT] [-g DBGEN_DIR]
                     [-s SCALE] [-n NUM_STREAMS] [-b] [-r]
                     {prepare,load,query}
tpch_pgsql.py: error: the following arguments are required: phase


# Tarefa 4 – Identificação do Sistema

In [ ]:
!df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay         108G   38G   71G  35% /
tmpfs            64M     0   64M   0% /dev
shm             5.8G   16K  5.8G   1% /dev/shm
/dev/root       2.0G  1.2G  812M  59% /sbin/docker-init
tmpfs           6.4G   28K  6.4G   1% /var/colab
/dev/sda1        81G   41G   40G  51% /etc/hosts
tmpfs           6.4G     0  6.4G   0% /proc/acpi
tmpfs           6.4G     0  6.4G   0% /proc/scsi
tmpfs           6.4G     0  6.4G   0% /sys/firmware


In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 b

In [ ]:
!cat /proc/meminfo

MemTotal:       13298580 kB
MemFree:         9616980 kB
MemAvailable:   12380664 kB
Buffers:          152788 kB
Cached:          2749236 kB
SwapCached:            0 kB
Active:          1157512 kB
Inactive:        2257128 kB
Active(anon):     469888 kB
Inactive(anon):     8372 kB
Active(file):     687624 kB
Inactive(file):  2248756 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:             22440 kB
Writeback:             0 kB
AnonPages:        512428 kB
Mapped:           268844 kB
Shmem:             12824 kB
KReclaimable:     137512 kB
Slab:             182480 kB
SReclaimable:     137512 kB
SUnreclaim:        44968 kB
KernelStack:        5056 kB
PageTables:         9072 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     6649288 kB
Committed_AS:    3243272 kB
VmallocTotal:   34359738367 kB
VmallocUsed:        7428 kB
VmallocChunk:          0 kB
Percpu:          

In [ ]:
!df -h / | awk '{print $4}'

Avail
71G


In [ ]:
!lscpu | grep cache

L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            56320K


In [ ]:
!wmic bios get serialnumber

/bin/bash: wmic: command not found


In [ ]:
cat /etc/os-release

NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic


In [ ]:
!uname -r

5.4.188+


Pela execução dos comandos acima, conseguimos as informações da máquina necessárias:

Hardware:
- Marca: Intel
- Modelo: 79 (Intel(R) Xeon(R) CPU @ 2.20GHz)
- Tipo de processador: 64 bits
- Quantidade de memoria ram: 13298580 kB
- Tamanho do disco: 108 GB (71GB disponíveis)
- Caches existentes: L1 32k, L2 256K, L3 56320K

Sistema operacional:
- Distribuição usada: Ubuntu
- Versão do sistema: 18.04.6 LTS
- Versão do Kernel: 5.4.188+

# Tarefa 6

Organização **PostgreSQL**: 
 
No PostgreSQL, os dados armazenados são referenciadaos no "PGDATA" que é o diretório de dados. Podem existir na mesma máquina vários agrupamentos, gerenciados por diferentes postmaster. 
Um local comum para PGDATA é /var/lib/pgsql/data. 
 
O Diretório PGDATA contém vários subdiretórios, dentre eles: 
 
PG_VERSION: Arquivo contendo o número de versão principal do PostgreSQL 
base: Subdiretório contendo subdiretórios por banco de dados"diário de reversão 
global: Subdiretório contendo tabelas para todo o agrupamento, como pg_database 
pg_clog: Subdiretório contendo dados sobre status de efetivação de transação 
pg_subtrans: Subdiretório contendo dados sobre status de subtransação 
 
Para cada banco de dados no cluster há um subdiretório dentro de PGDATA /base. Este subdiretório é o local padrão para os arquivos do banco de dados;  
 
Cada tabela e índice são armazenados em um arquivo separado. Para relações comuns, esses arquivos são nomeados de acordo com o número do nó de arquivo da tabela ou índice , que pode ser encontrado em pg_class . relfilenode .

Organização **SQLite** 
 
No SQLite os dados  em geral ficam em um unico arquivo, que é denominado "arquivo de banco de dados principal". Quando ocorre uma transação, 
as informações adicionais ficam armazenadas em um outro local. Este é denominado "diário de reversão".  
 
Se o aplicativo travar antes da conclusão da transação, o diário de reversão conterá as informações necessárias para restaurar o arquivo de banco de dados principal para um estado consistente. 
 
O arquivo de banco de dados principal consiste em uma ou mais páginas. O tamanho de uma página é uma potência de dois entre 512 e 65536. Todas as páginas dentro do mesmo banco de dados são do mesmo tamanho.

# Tarefa 7

<b> Ext2 </b>:
- Limite do tamanho do sistema de arquivo: 2TB até 32TB (dependendo do
tamanho do bloco, sendo este podendo variar de 1KB até 8KB)
- Não suporta journaling
- Utiliza tabela como estrutura de dados para conteúdos de diretório
- Alocação de arquivos usando bitmap e tabela para metadados
- Sistemas operacionais suportados: Linux, BSD, Windows (através de IFS)

<b> Ext3 </b>:
- Limite do tamanho do sistema de arquivo: 4TB até 32TB
- Suporta journaling
- Utiliza hashed B tree como estrutura de dados para conteúdos de diretório
- Alocação de arquivos usando bitmap e tabela para metadados
- Sistemas operacionais suportados: Linux, BSD, Windows e macOS
(quando utilizados por um IFS)

<b> ReiserFS </b>:
- Limite do tamanho do sistema de arquivo: 16TB
- Suporta journaling
- Utiliza B+ tree como estrutura de dados para conteúdos de diretório
- Alocação de arquivos usando bitmap
- Sistemas operacionais suportados: Linux, ReactOS

<b> XFS </b>:
- Limite do tamanho do sistema de arquivos: 8 Exbibytes
- Suporta journaling
- Utiliza B+ tree como estrutura de dados para conteúdos de diretório
- Alocação de arquivos usando B+ tree
- Sistema operacionais suportados: IRIX, Linux, FreeBSD

# Tarefa 10 -  Preparação de Tabela Exemplo

In [ ]:
%%sql
DROP TABLE IF EXISTS t;

CREATE TABLE t(k serial PRIMARY KEY, v integer);

INSERT INTO t(v) 
SELECT trunc(random() * 10)  FROM generate_series(1,100000);

 * postgresql://postgres:***@localhost:5432/sampledb
Done.
Done.
100000 rows affected.


[]

In [ ]:
%%sql

select *
FROM t 
ORDER BY k
fetch first 10 rows only

 * postgresql://postgres:***@localhost:5432/sampledb
10 rows affected.


[(1, 4),
 (2, 0),
 (3, 0),
 (4, 8),
 (5, 7),
 (6, 6),
 (7, 6),
 (8, 5),
 (9, 9),
 (10, 4)]

# Tarefa 11 - Páginas criadas

In [ ]:
%%sql
SELECT relname, relpages, reltuples 
FROM pg_class 
WHERE relname='t';

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.


[('t', 443, 100000.0)]

# Tarefa 12 - Blocos

In [ ]:
%%sql
SELECT pg_sleep(1); 
SELECT * FROM pg_stats WHERE tablename='t';
SELECT pg_stat_reset();

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.
0 rows affected.
1 rows affected.


[('',)]

# Tarefa 13 – Índice

Criação do índice

In [ ]:
%%sql
create index v_index on t(v);

 * postgresql://postgres:***@localhost:5432/sampledb
Done.


[]

**Qual o tempo gasto para realizar uma consulta para um valor (tendo a tabela 100000 tuplas)?**

In [ ]:
%%sql
EXPLAIN (ANALYZE, TIMING OFF)
SELECT '1' 
FROM t;

 * postgresql://postgres:***@localhost:5432/sampledb
3 rows affected.


[('Seq Scan on t  (cost=0.00..1443.00 rows=100000 width=32) (actual rows=100000 loops=1)',),
 ('Planning time: 0.069 ms',),
 ('Execution time: 16.915 ms',)]

Recriando index

In [ ]:
%%sql
DROP INDEX v_index;
create index v_index on t(v);

 * postgresql://postgres:***@localhost:5432/sampledb
Done.


[]

Dropando a tabela t

In [ ]:
%%sql
DROP TABLE t;

 * postgresql://postgres:***@localhost:5432/sampledb
Done.


[]

Recriando a nova tabela t

In [ ]:
%%sql
DROP TABLE IF EXISTS t;

CREATE TABLE t(k serial PRIMARY KEY, v integer);

INSERT INTO t(v) 
SELECT trunc(random() * 10)  FROM generate_series(1,10000000);

create index v_index on t(v);

 * postgresql://postgres:***@localhost:5432/sampledb
Done.
Done.
10000000 rows affected.
Done.


[]

**Qual o tempo gasto para realizar uma consulta para um valor (tendo a tabela 10000000 tuplas)?**

In [ ]:
%%sql
EXPLAIN (ANALYZE, TIMING OFF)
SELECT '1' 
FROM t;

 * postgresql://postgres:***@localhost:5432/sampledb
3 rows affected.


[('Seq Scan on t  (cost=0.00..144248.00 rows=10000000 width=32) (actual rows=10000000 loops=1)',),
 ('Planning time: 0.736 ms',),
 ('Execution time: 1343.109 ms',)]

Recriando índice

In [ ]:
%%sql
DROP INDEX v_index;
create index v_index on t(v);

 * postgresql://postgres:***@localhost:5432/sampledb
Done.
Done.


[]

Resultado:

Consulta com 100000 elementos: 16.915ms

Consulta com 10000000 elementos: 1343.109ms

# Tarefa 14 - Fill factor

In [ ]:
%%sql
ALTER TABLE t SET (fillfactor =100);
VACUUM FULL t;

 * postgresql://postgres:***@localhost:5432/sampledb
Done.
Done.


[]

In [ ]:
%%sql
EXPLAIN (ANALYZE, TIMING OFF)
SELECT '1' 
FROM t;

 * postgresql://postgres:***@localhost:5432/sampledb
3 rows affected.


[('Seq Scan on t  (cost=0.00..1443.00 rows=100000 width=32) (actual rows=100000 loops=1)',),
 ('Planning time: 0.251 ms',),
 ('Execution time: 16.597 ms',)]

Resultados das consultas de um elemento:

<br>

Fill factor 60:

100000 elementos: 24.761 ms

10000000 elementos: 1410.132 ms

<br>

Fill factor 80:

100000 elementos: 25.930 ms

10000000 elementos: 1329.337 ms

<br>

Fill factor 90:

100000 elementos: 18.353 ms

10000000 elementos: 1329.939 ms

<br>

Fill factor 100:

100000 elementos: 16.597 ms

10000000 elementos: 1339.496 ms

# Tarefa 16 - Utilize índices com ordem DESC

In [ ]:
%%sql
DROP TABLE IF EXISTS t2;
CREATE TABLE t2 (k serial PRIMARY KEY, v integer);
create index v_index on t2(v DESC NULLS FIRST);

INSERT INTO t2(v)
SELECT trunc(random() * 10)  FROM generate_series(1,100000);

Done.
Done.
Done.
100000 rows affected.


[]

Refazendo a tarefa 12: Paginas criadas

In [ ]:
%%sql
SELECT relname, relpages, reltuples FROM pg_class WHERE relname='t2';

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.


[('t2', 443, 100000.0)]

Refazendo Tarefa 13: Blocos

In [ ]:
%%sql
SELECT pg_sleep(1);
SELECT * FROM pg_stats WHERE tablename='t2';
SELECT pg_stat_reset();

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.
2 rows affected.
1 rows affected.


[('',)]

Refazendo Tarefa 14: Índice

In [ ]:
%%sql
EXPLAIN (ANALYZE, TIMING OFF)
SELECT '1' 
FROM t2;

 * postgresql://postgres:***@localhost:5432/sampledb
3 rows affected.


[('Seq Scan on t2  (cost=0.00..144248.48 rows=10000048 width=32) (actual rows=10000000 loops=1)',),
 ('Planning time: 0.165 ms',),
 ('Execution time: 1878.019 ms',)]

**Resultado:**

Consulta com 100000 elementos: 16.167 ms

Consulta com 10000000 elementos: 1878.019 ms

# Tarefa 17 - Preparação e Verificação do Ambiente

Listando todos os índices

In [ ]:
%%sql

SELECT
    tablename,
    indexname,
    indexdef
FROM
    pg_indexes
WHERE
    schemaname = 'public'
ORDER BY
    tablename,
    indexname;

 * postgresql://postgres:***@localhost:5432/sampledb
9 rows affected.


[('actor', 'actor_id', 'CREATE INDEX actor_id ON public.actor USING btree (id)'),
 ('casting', 'casting_actor', 'CREATE INDEX casting_actor ON public.casting USING btree (actorid)'),
 ('casting', 'casting_movie', 'CREATE INDEX casting_movie ON public.casting USING btree (movieid)'),
 ('casting', 'casting_ord', 'CREATE INDEX casting_ord ON public.casting USING btree (ord)'),
 ('movie', 'movie_key', 'CREATE UNIQUE INDEX movie_key ON public.movie USING btree (id)'),
 ('movie', 'movie_title', 'CREATE INDEX movie_title ON public.movie USING btree (title)'),
 ('movie', 'movie_votes', 'CREATE INDEX movie_votes ON public.movie USING btree (votes)'),
 ('t2', 't2_pkey', 'CREATE UNIQUE INDEX t2_pkey ON public.t2 USING btree (k)'),
 ('t2', 'v_index', 'CREATE INDEX v_index ON public.t2 USING btree (v DESC)')]

Obtendo informações adicionais dos índices

In [ ]:
%%sql
SELECT * FROM pgstatindex('movie_key');

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.


[(2, 1, 57344, 3, 1, 5, 0, 0, 90.73, 0.0)]

In [ ]:
%%sql
SELECT * FROM pgstatindex('movie_title');

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.


[(2, 1, 90112, 3, 1, 9, 0, 0, 80.36, 44.44)]

In [ ]:
%%sql
SELECT * FROM pgstatindex('movie_votes');

 * postgresql://postgres:***@localhost:5432/sampledb
1 rows affected.


[(2, 1, 90112, 3, 1, 9, 0, 0, 50.53, 77.78)]

Não consegui informações adicionais sobre os outros níveis de árvore

# Tarefa 18 - Consultas por intervalo e índices secundários

questao **a)**

In [ ]:
%%sql
EXPLAIN
select votes
FROM movie
fetch first 10 rows only

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Limit  (cost=0.00..0.18 rows=10 width=4)',),
 ('  ->  Seq Scan on movie  (cost=0.00..33.44 rows=1844 width=4)',)]

questao **b)**

In [ ]:
%%sql
EXPLAIN
select votes
FROM movie
fetch first 1700 rows only

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Limit  (cost=0.00..30.83 rows=1700 width=4)',),
 ('  ->  Seq Scan on movie  (cost=0.00..33.44 rows=1844 width=4)',)]

questao **c)**

O índice sobre VOTES não é sempre usado nas consultas porque é uma decisão do otimizador de consultas. Existe um equilíbrio entre os ganhos de desempenho por consultas executadas usando esses índices, e também com o tempo necessário para manter e atualizar esses índices.

Nas duas consultas executadas acima, não está sendo utilizado o index.

# Tarefa 19. Comparações de operadores de agregação

questao **a)**

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE votes >= (SELECT MAX(votes) FROM movie);

 * postgresql://postgres:***@localhost:5432/sampledb
8 rows affected.


[('Index Scan using movie_votes on movie  (cost=0.62..35.38 rows=615 width=16)',),
 ('  Index Cond: (votes >= $1)',),
 ('  InitPlan 2 (returns $1)',),
 ('    ->  Result  (cost=0.33..0.34 rows=1 width=4)',),
 ('          InitPlan 1 (returns $0)',),
 ('            ->  Limit  (cost=0.28..0.33 rows=1 width=4)',),
 ('                  ->  Index Only Scan Backward using movie_votes on movie movie_1  (cost=0.28..94.55 rows=1844 width=4)',),
 ('                        Index Cond: (votes IS NOT NULL)',)]

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE votes >= ALL (SELECT votes FROM movie) ;

 * postgresql://postgres:***@localhost:5432/sampledb
5 rows affected.


[('Seq Scan on movie  (cost=0.00..43620.99 rows=922 width=16)',),
 ('  Filter: (SubPlan 1)',),
 ('  SubPlan 1',),
 ('    ->  Materialize  (cost=0.00..42.66 rows=1844 width=4)',),
 ('          ->  Seq Scan on movie movie_1  (cost=0.00..33.44 rows=1844 width=4)',)]

questao **b)**

**Sim! Existe diferença entre as duas consultas**

A primeira consulta foi mais eficiente!

O tempo de planejamento da primeira foi maior, mas seu tempo de execução foi menor.

Isso se deve, em parte, ao fato de a primeira consulta estar usando um index e a segunda consulta não.

# Tarefa 20. Consultas com Junção e Seleção

questao **a)**

In [ ]:
%%sql
EXPLAIN ANALYSE
SELECT title FROM movie WHERE votes > (SELECT votes FROM movie WHERE title = 'Star Wars');

 * postgresql://postgres:***@localhost:5432/sampledb
7 rows affected.


[('Index Scan using movie_votes on movie  (cost=8.57..43.34 rows=615 width=16) (actual time=0.046..0.047 rows=0 loops=1)',),
 ('  Index Cond: (votes > $0)',),
 ('  InitPlan 1 (returns $0)',),
 ('    ->  Index Scan using movie_title on movie movie_1  (cost=0.28..8.29 rows=1 width=4) (actual time=0.034..0.035 rows=1 loops=1)',),
 ("          Index Cond: ((title)::text = 'Star Wars'::text)",),
 ('Planning time: 0.166 ms',),
 ('Execution time: 0.079 ms',)]

In [ ]:
%%sql
EXPLAIN ANALYSE
SELECT m1.title FROM movie m1, movie m2 WHERE m1.votes > m2.votes AND m2.title = 'Star Wars'; 

 * postgresql://postgres:***@localhost:5432/sampledb
7 rows affected.


[('Nested Loop  (cost=0.56..49.49 rows=615 width=16) (actual time=0.047..0.047 rows=0 loops=1)',),
 ('  ->  Index Scan using movie_title on movie m2  (cost=0.28..8.29 rows=1 width=4) (actual time=0.034..0.035 rows=1 loops=1)',),
 ("        Index Cond: ((title)::text = 'Star Wars'::text)",),
 ('  ->  Index Scan using movie_votes on movie m1  (cost=0.28..35.04 rows=615 width=20) (actual time=0.009..0.009 rows=0 loops=1)',),
 ('        Index Cond: (votes > m2.votes)',),
 ('Planning time: 0.220 ms',),
 ('Execution time: 0.087 ms',)]

questao **b)**

**Sim! Existe diferença entre as duas consultas**

A primeira consulta foi mais eficiente que a segunda.

Pois, as duas consultas estão utilizando índices, mas a segunda consulta também utiliza um nested loop, onde para cara valor da tabela é necessário fazer uma busca de igualdade na outra tabela.

# Tarefa 21.  Casamento de Strings e Índices

questao **a)**

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE title LIKE 'I%';

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16)',),
 ("  Filter: ((title)::text ~~ 'I%'::text)",)]

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE substr(title, 1, 1) = 'I'; 

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Seq Scan on movie  (cost=0.00..42.66 rows=9 width=16)',),
 ("  Filter: (substr((title)::text, 1, 1) = 'I'::text)",)]

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE title LIKE '%A'; 

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Seq Scan on movie  (cost=0.00..38.05 rows=18 width=16)',),
 ("  Filter: ((title)::text ~~ '%A'::text)",)]

questao **b)**

A terceira consulta apresenta o menor custo!

Pois o filtro resultou em um número menor de linhas para as operações subsequentes.

questao **c)**

Não! O índice sobre TITLE não foi usado para nenhuma das consultas

Provavelmente por estarmos fazendo operações de filtro com a tabela inteira, um índice não traria muita vantagem nesse tipo de operação

# Tarefa 22. Verificação da hipótese de distribuição uniforme na estimativa de seletividade


questao **a)**

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE votes < 1000; 

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Index Scan using movie_votes on movie  (cost=0.28..20.04 rows=329 width=16)',),
 ('  Index Cond: (votes < 1000)',)]

In [ ]:
%%sql
EXPLAIN
SELECT title FROM movie WHERE votes > 40000 

 * postgresql://postgres:***@localhost:5432/sampledb
2 rows affected.


[('Index Scan using movie_votes on movie  (cost=0.28..8.42 rows=8 width=16)',),
 ('  Index Cond: (votes > 40000)',)]

As duas consultas utilizaram os índices, mas a segunda consulta foi mais eficiente, pois existem bem menos tuplas com esse filtro

questao **b)**

Tuplas selecionadas:

Primeira consulta: **326**

Segunda consulta: **4**

A primeira consulta tem mais seletividade